File: tests/journal.lib/chronicler.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the journal
#include <p2/journal.h>
// support
#include <cassert>


// alias
using chronicler_t = pyre::journal::chronicler_t;


// verify that the initial chronicler global state is as expected
int main() {
    // verify that the default verbosity is at level 1
    assert (chronicler_t::verbosity() == 1);

    // get the global metadata map
    chronicler_t::metadata_type & globals = chronicler_t::globals();

    // there should be only one setting for now
    assert (globals.size() == 1);
    // verify the known contents
    assert (globals.at("application") == "journal");

    // get the default device
    auto device_ptr = chronicler_t::device();
    // by default, it is the console; verify that it is not an empty pointer
    assert(device_ptr);
    // and that it's the console
    assert ((*device_ptr).name() == "cout");

    // all done
    return 0;
}


// end of file

File: pkg/j2/Chronicler.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# for my metaclass
import p2


# the singleton that owns the global state
class Chronicler(metaclass=p2.patterns.singleton):
    """
    The manager of the journal global state
    """


    # public data
    meta = None
    device = None
    verbosity = 1


    # metamethods
    def __init__(self, verbosity=verbosity, device=device, meta=meta, **kwds):
        # chain up
        super().__init__(**kwds)

        # the default verbosity
        self.verbosity = verbosity
        # the default device
        self.device = device
        # the global metadata
        self.meta = meta if meta is not None else {
            "application": "journal",  # this key is required; applications should override
            }

        # all done
        return


# end of file
